src/System.Diagnostics.Debug/tests/DebugTests.cs

```diff
@@ -200,64 +196,107 @@ static void VerifyAssert(Action test, params string[] expectedOutputStrings)
-
-            var originalShowDialogHook = showDialogHook.GetValue(null);
-            showDialogHook.SetValue(null, new Action<string, string, string, string>(WriteLogger.s_instance.ShowDialog));
+            Debug.SetProvider(WriteLogger.s_writeLoggerInstance);
 
             try
             {
-                WriteLogger.s_instance.Clear();
+                WriteLogger.s_writeLoggerInstance.Clear();
                 test();
 #if DEBUG
                 for (int i = 0; i < expectedOutputStrings.Length; i++)
                 {
-                    Assert.Contains(expectedOutputStrings[i], WriteLogger.s_instance.LoggedOutput);
-                    Assert.Contains(expectedOutputStrings[i], WriteLogger.s_instance.AssertUIOutput);
+                    Assert.Contains(expectedOutputStrings[i], WriteLogger.s_writeLoggerInstance.LoggedOutput);
+                    Assert.Contains(expectedOutputStrings[i], WriteLogger.s_writeLoggerInstance.AssertUIOutput);
                 }
 #else
-                Assert.Equal(string.Empty, WriteLogger.s_instance.LoggedOutput);
-                Assert.Equal(string.Empty, WriteLogger.s_instance.AssertUIOutput);
+                Assert.Equal(string.Empty, WriteLogger.s_writeLoggerInstance.LoggedOutput);
+                Assert.Equal(string.Empty, WriteLogger.s_writeLoggerInstance.AssertUIOutput);
 #endif
 
             }
             finally
             {
-                writeCoreHook.SetValue(null, originalWriteCoreHook);
-                showDialogHook.SetValue(null, originalShowDialogHook);
+                Debug.SetProvider(s_basicProvider);
             }
         }
 
-        private class WriteLogger
+        protected class LogProvider : Debug.DebugProvider
         {
-            public static readonly WriteLogger s_instance = new WriteLogger();
+            public static readonly LogProvider s_logProviderInstance = new LogProvider();
 
-            private WriteLogger() { }
+            protected LogProvider() { }
 
-            public string LoggedOutput { get; private set; }
+            public string LoggedOutput { get; protected set; }
 
-            public string AssertUIOutput { get; private set; }
-
-            public void Clear()
+            public virtual void Clear()
             {
                 LoggedOutput = string.Empty;
-                AssertUIOutput = string.Empty;
             }
 
-            public void ShowDialog(string stackTrace, string message, string detailMessage, string errorSource)
+            public override void Write(string message)
             {
-                AssertUIOutput += stackTrace + message + detailMessage + errorSource;
+                OldWrite(message);
             }
 
-            public void WriteCore(string message)
+            private void WriteCore(string message)
             {
                 Assert.NotNull(message);
                 LoggedOutput += message;
             }
+            
+            private static bool s_needIndent;
```

Comment:
I added the lines between 247-280 to confirm when LogProvider.Write on line 238 does the same logic as DebugInternal.Write in Debug.cs then the test would (add did) behave as before. The only difference is the WriteCore on line 241 here has a different logic than that in DebugInternal.

If we want to keep the indent tests in DebugTests these lines need to stay.